File: blue-serial/src/main/java/io/github/chayanforyou/blueserial/BluetoothConnectionBase.java
package io.github.chayanforyou.blueserial;

public abstract class BluetoothConnectionBase implements BluetoothConnection {

    /// Callback for reading data.
    public interface OnReadCallback {
        void onRead(byte[] data);
    }

    /// Callback for disconnection.
    public interface OnDisconnectedCallback {
        void onDisconnected(boolean byRemote);
    }

    final OnReadCallback onReadCallback;
    final OnDisconnectedCallback onDisconnectedCallback;

    public BluetoothConnectionBase(OnReadCallback onReadCallback, OnDisconnectedCallback onDisconnectedCallback) {
        this.onReadCallback = onReadCallback;
        this.onDisconnectedCallback = onDisconnectedCallback;
    }

    public void onRead(byte[] data) {
        onReadCallback.onRead(data);
    }

    public void onDisconnected(boolean byRemote) {
        onDisconnectedCallback.onDisconnected(byRemote);
    }
}

File: blue-serial/src/main/java/io/github/chayanforyou/blueserial/BluetoothConnectionBase.kt
package io.github.chayanforyou.blueserial

abstract class BluetoothConnectionBase(
    protected val onReadCallback: OnReadCallback,
    protected val onDisconnectedCallback: OnDisconnectedCallback
) : BluetoothConnection {

    /// Callback for reading data.
    fun interface OnReadCallback {
        fun onRead(data: ByteArray)
    }

    /// Callback for disconnection.
    fun interface OnDisconnectedCallback {
        fun onDisconnected(byRemote: Boolean)
    }

    open fun onRead(data: ByteArray) {
        onReadCallback.onRead(data)
    }

    open fun onDisconnected(byRemote: Boolean) {
        onDisconnectedCallback.onDisconnected(byRemote)
    }
}
